### PR TITLE
add readme for librustdoc

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
In the same vein as the other compiler-library readmes, here's one for rustdoc! It's mainly a "how does rustdoc even" blog-post-style writeup, but i wanted to have something in-repo so people could get a sense of what bits did what.

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
@GuillaumeGomez @steveklabnik Did i miss anything glaringly obvious here? This file kinda sat half-written on my server for a few months, so at some point i just wanted to get *something* written up so we could have it online.

**Comment by mark-i-m** [comment]
@QuietMisdreavus Could you add/adapt the same for [rustc-guide](https://github.com/rust-lang-nursery/rustc-guide)?

Also, I'm guessing this is for current rustdoc, not https://github.com/steveklabnik/rustdoc?

**Comment by QuietMisdreavus** [comment]
@mark-i-m Yeah, i can add this when it gets approved up here. Don't want to go through the same review nits more than once, haha.

Also yes, since the readme is sitting in `src/librustdoc`, it's describing that code. I wanted it to be like the README.md in `src/librustc`, where it can be seen alongside the code when browsing on GitHub.

**Comment by GuillaumeGomez** [comment]
Sounds good to me but I'd prefer an english native speaker to give a look before r+.

File: src/librustdoc/README.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  * Add `src/libtest` to be able to use `rustdoc --test`.
+  * If you've used `rustup toolchain link local /path/to/build/$TARGET/stage1` previously, then
+    after the previous build command, `cargo +local doc` will Just Work.
+* Use `x.py doc --stage 1 src/libstd` to use this rustdoc to generate the standard library docs.
+  * The completed docs will be available in `build/$TARGET/doc/std`, though the bundle is meant to
+    be used as though you would copy out the `doc` folder to a web server, since that's where the
+    CSS/JS and landing page are.
+* Most of the HTML printing code is in `html/format.rs` and `html/render.rs`. It's in a bunch of
+  `fmt::Display` implementations and supplementary functions.
+* The types that operates on are defined in `clean/mod.rs`, right next to the custom `Clean` trait
```
Comment (mark-i-m) [inline src/librustdoc/README.md]: "The types that operates on"

File: src/librustdoc/README.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+You see, I actually lied a little earlier: There's another AST transformation that happens before
+the events in `clean/mod.rs`.  In `visit_ast.rs` is the type `RustdocVisitor`, which *actually*
+crawls a `hir::Crate` to get the first intermediate representation, defined in `doctree.rs`. This
+pass is mainly to get a few intermediate wrappers around the HIR types and to process visibility
+and inlining. This is where `#[doc(inline)]`, `#[doc(no_inline)]`, and `#[doc(hidden)]` are
+processed, as well as the logic for whether a `pub use` should get the full page or a "Reexport"
+line in the module page.
+
+Back in `clean/mod.rs`, the other major thing that happens here is the special collection of doc
+comments (and basic `#[doc=""]` attributes) into a separate field in the `Attributes` struct from
```
Comment (mark-i-m) [inline src/librustdoc/README.md]: This sentence is a bit hard to follow.

File: src/librustdoc/README.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+which describe the publicly-documentable items in the target crate.
+
+### Hot potato
+
+Before moving on to the next major step, a few important "passes" occur over the documentation.
+These do things like combine the separate "attributes" into a single string and strip leading
+whitespace to make the document easier on the markdown parser, or drop items that are not public or
+deliberately hidden with `#[doc(hidden)]`. These are all implemented in the `passes/` directory, one
+file per pass. By default, all of these passes are run on a crate, but the ones regarding dropping
+private/hidden items can be bypassed by passing `--document-private-items` to rustdoc.
```
Comment (mark-i-m) [inline src/librustdoc/README.md]: It would be good to note where the passes are called from so that adding a new pass is easier...

**Review by mark-i-m — CHANGES_REQUESTED** [review]
Overall, LGTM. I left a few comments about minor nits...

File: src/librustdoc/README.md (hunk 1 above)
Comment (QuietMisdreavus) [reply src/librustdoc/README.md]: How about:

> The types that got `Display` impls above are defined in...

File: src/librustdoc/README.md (hunk 2 above)
Comment (QuietMisdreavus) [reply src/librustdoc/README.md]: How about:

> The other major thing that happens in `clean/mod.rs` is the collection of doc comments and `#[doc=""]` attributes into a separate field of the `Attributes` struct, present on anything that gets hand-written documentation. This makes it easier to collect this documentation later in the process.
>
> The primary output of this process is a `clean::Crate` with a tree of `Item`s which describe the publicly-documentable items in the target crate.

These statements weren't that related anyway, so breaking that last one out into its own paragraph makes that explicit.

File: src/librustdoc/README.md (hunk 3 above)
Comment (QuietMisdreavus) [reply src/librustdoc/README.md]: I avoided mentioning the passes explicitly, since we kinda wanted to abandon the concept, but i can leave a note. It's awkward and wrapped up in the plugin interface that we've explicitly deprecated.

File: src/librustdoc/README.md (hunk 1 above)
Comment (mark-i-m) [reply src/librustdoc/README.md]: Ah, I see. Yeah, that's a lot better.

File: src/librustdoc/README.md (hunk 3 above)
Comment (mark-i-m) [reply src/librustdoc/README.md]: Ah, I see... In that case, maybe it would be better to _not_ document that so it is harder to add another pass.

Comment (QuietMisdreavus) [reply src/librustdoc/README.md]: In other words, take out this section entirely? The passes do some fairly important things, so i would be remiss to leave it out. The entry to the passes isn't necessarily *difficult*, just awkward. Technically each "pass" is a plugin, loaded up at the same time as any dynamically-loaded plugin that could potentially have been given.

https://github.com/rust-lang/rust/blob/27a046e9338fb0455c33b13e8fe28da78212dedc/src/librustdoc/lib.rs#L583-L612

Comment (mark-i-m) [reply src/librustdoc/README.md]: No, I think we can just leave it as is. Those who need to deal with it can ask for further clarification when they cross that bridge.

**Review by mark-i-m — APPROVED** [review]

**Comment by mark-i-m** [comment]
@GuillaumeGomez r=english-speaker=me

**Comment by shepmaster** [comment]
Everyone likes progress, yeah?

@bors r=@GuillaumeGomez

**Comment by mark-i-m** [comment]
:tada: perhaps it's time to add it to the rustc guide?
